### PR TITLE
Export features

### PR DESCRIPTION
**Pull request by sjaanus**
1. Support file download
2. Add event for export
3. Segment support
4. Custom context fields support 
5. Tags support

### CODE REVIEW

File: src/lib/services/export-import-service.ts [hunk 1]
```diff
@@ -95,24 +93,70 @@ export default class ExportImportService {
...
+        const [
+            features,
+            featureEnvironments,
+            featureStrategies,
+            strategySegments,
+            contextFields,
+            featureTags,
+        ] = await Promise.all([
+            this.toggleStore.getAllByNames(query.features),
+            await this.featureEnvironmentStore.getAll((builder) => {
```
Comment (sjaanus) [inline src/lib/services/export-import-service.ts:113]: I would love some feedback on this. Most of our `getAll` methods allows us build queries outside store like this.

**Pros:** Dont need to create a separate store function every time we want to run a different query on dataset.
**Cons:** We write Knex in service level.

File: src/lib/routes/admin-api/export-import.ts [hunk 2]
```diff
@@ -58,19 +60,27 @@ class ExportImportController extends Controller {
...
+        const data = await this.exportImportService.export(query, userName);
 
         this.openApiService.respondWithValidation(
             200,
             res,
             exportResultSchema.$id,
             serializeDates(data),
         );
+
+        const timestamp = formatDate(Date.now(), 'yyyy-MM-dd_HH-mm-ss');
```
Comment (kwasniew) [inline src/lib/routes/admin-api/export-import.ts:78]: I'd extract a method and maybe made it more testable - no hardcoded date

File: src/lib/services/export-import-service.ts [hunk 3]
```diff
@@ -95,24 +93,70 @@ export default class ExportImportService {
...
+        await this.eventStore.store({
+            type: FEATURES_EXPORTED,
+            createdBy: userName,
+            data: result,
+        });
+
+        return result;
+    }
+
+    addSegmentsToStrategies(
```
Comment (kwasniew) [inline src/lib/services/export-import-service.ts:148]: could we have ExportStore with a query that will avoid O(N2) problem?

File: src/test/e2e/api/admin/export-import.e2e.test.ts [hunk 4]
```diff
@@ -66,6 +98,13 @@ const createProject = async (project: string, environment: string) => {
             environment,
         })
         .expect(200);
+    // await db.stores.projectStore.addEnvironmentToProject(project, environment);
```
Comment (kwasniew) [inline src/test/e2e/api/admin/export-import.e2e.test.ts:101]: leftover?

File: src/test/e2e/api/admin/export-import.e2e.test.ts [hunk 5]
```diff
@@ -150,6 +191,106 @@ test('exports features', async () => {
...
+    const { body } = await app.request
+        .post('/api/admin/features-batch/export')
+        .send({
+            features: ['first_feature'],
+            environment: 'default',
+        })
+        .set('Content-Type', 'application/json')
+        .expect(200);
+
+    const { name, ...resultStrategy } = strategy;
```
Comment (kwasniew) [inline src/test/e2e/api/admin/export-import.e2e.test.ts:233]: I'd leave the name for the import to work

File: src/test/e2e/api/admin/export-import.e2e.test.ts [hunk 6]
```diff
@@ -150,6 +191,106 @@ test('exports features', async () => {
...
+                variants: [],
+            },
+        ],
+        contextFields: [context],
+    });
+});
+
+test('should export tags', async () => {
+    const featureName = 'first_feature';
+    await createProject('default', 'default');
```
Comment (kwasniew) [inline src/test/e2e/api/admin/export-import.e2e.test.ts:255]: I'd extract env variable and use env other than default e.g. production

**Review by kwasniew — APPROVED** [review]

**Pull request merged**